src/api_rework.jl

```diff
@@ -0,0 +1,116 @@
+# I am putting everything in here for now. The contents of this file should be moved around in the future.
+"""
+    findpeaks(x) -> NamedTuple
+
+Find the peaks in a vector `x`. 
+A `NamedTuple` is returned with the original vector 
+in the field `data`, and the indices of the peaks 
+in the field `inds`.
+
+This function serves as the entry-point for other 
+functions such as `peakproms!` and `peakwidths!`
+"""
+findpeaks(x::AbstractVector) = (data=x, inds=findmaxima(x)[1])
+
+"""
+    _filter_fields!(pks, new_inds)
+
+Internal function to mutate the vectors in all fields of `pks` 
+to remove any element with a corresponding `pks.inds` entry 
+not present in `new_inds`.
+"""
+function _filter_fields!(pks, new_inds)
+    # Add checks on `pks` to see if it has fields "data" and "inds"?
+    mask = pks.inds .∉ Ref(new_inds)
+    for field in filter(!=(:data), propertynames(pks)) # Do not want to mutate data vector
+        v_to_be_mutated = getfield(pks, field)
+        if length(v_to_be_mutated) != length(mask)
+            error("Internal error: The length of the vector in field `$field` is $(length(v_to_be_mutated)), but was expected to be $(length(mask))")
+        end
```

Comment:
This should be hoisted into a separate loop. We don't want to realize that some of the fields are inconsistent after we've already modified some (making any inconsistencies even worse). Check out [this line](https://github.com/halleysfifthinc/Peaks.jl/blob/8a415b5fed1db8982fb02551ed8adf88b5f8e173/src/peakheight.jl#L61) for an example.

Reply:
Good point. The function is now fixed.
<details>
<summary>New version, with new name</summary>

```julia
function filterpeaks!(pks, mask::BitVector)
    features_to_filter = (:indices, :proms, :heights, :widths, :edges)

    # Check lengths first to avoid a dimension mismatch 
    # after having filtered some features.
    for field in features_to_filter
        if length(mask) != length(getfield(pks, field))
            if length(mask) != length(getfield(pks, field))
            throw(DimensionMismatch(
            "Length of `mask` is ($(length(mask))), but the length 
            of `pks.$field` is $(length(getfield(pks, field))). 
            This means that the given mask can not be used to filter 
            the field `$field`." ))
        end
        end
    end

    for field in features_to_filter  # Only risk mutating fields added by this package
        hasfield(pks, field) || continue  # Do nothing if field is not present
        v_to_be_mutated = getfield(pks, field)
        deleteat!(v_to_be_mutated, mask)
    end
    return nothing
end
```